util/syspolicy/internal/metrics/metrics.go

```diff
@@ -0,0 +1,315 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package metrics provides logging and reporting for policy settings and scopes.
+package metrics
+
+import (
+	"strings"
+	"sync"
+
+	xmaps "golang.org/x/exp/maps"
+
+	"tailscale.com/syncs"
+	"tailscale.com/types/lazy"
+	"tailscale.com/util/clientmetric"
+	"tailscale.com/util/mak"
+	"tailscale.com/util/syspolicy/internal"
+	"tailscale.com/util/syspolicy/internal/loggerx"
+	"tailscale.com/util/syspolicy/setting"
+	"tailscale.com/util/testenv"
+)
+
+var lazyReportMetrics lazy.SyncValue[bool] // used as a test hook
+
+// ShouldReport reports whether metrics should be reported on the current environment.
+func ShouldReport() bool {
+	return lazyReportMetrics.Get(func() bool {
+		// macOS, iOS and tvOS create their own metrics,
+		// and we don't have syspolicy on any other platforms.
+		return setting.PlatformList{"android", "windows"}.HasCurrent()
+	})
```

Comment:
We plan to use it across all platforms, including iOS/macOS, which currently report their own metrics, and Unix-like platforms, which lack `syspolicy` support today. At a minimum, I think we can implement an [envvar-based policy store](https://github.com/tailscale/tailscale/issues/13193#issuecomment-2298084842) and/or a config-like policy store. Additionally, there are discussions about allowing `syspolicy` configuration via the control plane, which would serve as a cross-platform policy source.

That being said, I definitely want to exclude test-related code from production builds, especially on iOS. My original plan was to move certain types and functions to `_notios.go` files in a follow-up PR.
However, I'm very interested in the following option and wasn't aware that our CI supports it.
> For testing, we could also use a build tag to include the full version for tests.

Do you know of a good example where we use this approach?

Reply:
no, but you just have to do something like `//go:build windows || android || test` -- the convention of putting tests in `foo_test.go` is using the compiler's implicit filename-based build tag logic to avoid including the test files if not running a test.

Reply:
Anyway to address the bigger response about planning to use it for all platforms, that's fine, but the current iteration isn't for all platforms, so it feels like it might be nice to avoid including it non the currently non-supported ones? If the plan isn't for things to stay that way, then don't put a ton of effort into it, but I don't think it should be super hard.

Anyway, assuming it's not putting us over any size thresholds we're worried about, I guess it's an optional suggestion.

Reply:
I see. But the `_test` suffix works differently and does not set the `test` build tag.
Instead, it makes the test code a part of a test package.

So this:
```
//go:build windows || android || test
```
will work if you run the tests on Windows (or Android if that's possible).
But it won't work when running tests on Linux unless we _explicitly_ set the `test` tag when we run the tests.
Hence the question about the CI support.

Reply:
That said, I agree with your second point and am happy to hide what's possible under the `!ios` tag (or its file suffix equivalent).